vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml

```diff
@@ -352,52 +352,61 @@
+  - name: Get cluster uid
+    kubernetes.core.k8s_info:
+      api_version: cluster.x-k8s.io/v1alpha3
+      kind: Cluster
+      name: "{{ CLUSTER_NAME }}"
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: clusters
 
-  - name: Create controlplane and worker Metal3MachineTemplates
+  - name: Create controlplane Metal3MachineTemplates
     kubernetes.core.k8s:
       state: present
-      src: /tmp/cp_new_image.yaml
+      template: Metal3MachineTemplate.yml
       kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    vars:
+      CLUSTER_UID: "{{ clusters.resources[0].metadata.uid }}"
+      M3MT_NAME: "{{CLUSTER_NAME}}-new-controlplane-image"
+      DATA_TEMPLATE_NAME: "{{CLUSTER_NAME}}-controlplane-template"
 
-  - name: Create controlplane and worker Metal3MachineTemplates
+  - name: Create worker Metal3MachineTemplates
     kubernetes.core.k8s:
       state: present
-      src:  /tmp/wr_new_image.yaml
+      template: Metal3MachineTemplate.yml
       kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    vars:
+      CLUSTER_UID: "{{ clusters.resources[0].metadata.uid }}"
+      M3MT_NAME: "{{CLUSTER_NAME}}-new-workers-image"
+      DATA_TEMPLATE_NAME: "{{CLUSTER_NAME}}-workers-template"
 
   - name: Update boot-disk and kubernetes versions of controlplane nodes
-    shell: |
-            kubectl get kubeadmcontrolplane -n {{NAMESPACE}} {{ CLUSTER_NAME }} -o json |
-            jq '.spec.infrastructureTemplate.name="{{CLUSTER_NAME}}-new-controlplane-image" |
-               .spec.version="{{UPGRADED_K8S_VERSION}}"'|
-                kubectl apply -f-
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    kubernetes.core.k8s:
+      api_version: controlplane.cluster.x-k8s.io/v1alpha3
+      kind: KubeadmControlPlane
+      name: "{{ CLUSTER_NAME }}"
+      namespace: "{{ NAMESPACE }}"
+      resource_definition:
+        spec:
+          version: "{{UPGRADED_K8S_VERSION}}"
+          infrastructureTemplate:
+            name: "{{CLUSTER_NAME}}-new-controlplane-image"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
 
+  # This is tricky. The control plane nodes are being replaced one by one.
+  # This naturally causes some interruptions in the communication with the API.
+  # Unfortunately it seems like the k8s_info module can get stuck in these
+  # situations (i.e. it doesn't fail, it just hangs).
+  # Ref https://github.com/ansible/ansible/issues/30411
+  #
+  # Attempted workarounds:
+  # - Use timeout (https://docs.ansible.com/ansible/latest/reference_appendices/playbooks_keywords.html#term-timeout)
+  #   This prevents hanging but fails the whole task instead of retrying.
+  #
+  # Issues that may be related
+  # - https://github.com/kubernetes-client/python/issues/1402
+  # - https://github.com/kubernetes-client/python/issues/602
   - name: Verify that controlplane nodes using the new node image
```

Comment:
# This is tricky. The control plane nodes are being replaced one by one.
  # This naturally causes some interruptions in the communication with the API.
  # Unfortunately it seems like the k8s_info module can get stuck in these
  # situations (i.e. it doesn't fail, it just hangs).
  # Ref https://github.com/ansible/ansible/issues/30411
  #
  # Attempted workarounds:
  # - Use timeout (https://docs.ansible.com/ansible/latest/reference_appendices/playbooks_keywords.html#term-timeout)
  #   This prevents hanging but fails the whole task instead of retrying.
  #
  # Issues that may be related
  # - https://github.com/kubernetes-client/python/issues/1402
  # - https://github.com/kubernetes-client/python/issues/602